### PR TITLE
Command line arguments

### PR DESCRIPTION
**Pull request by pieter-edelman-nictiz**
#### Brief overview of PR changes/additions
Prototype for starting Hammer with commend line arguments. Currently implemented are an `-s/--scope` option to set the scope directory and a positional argument to load an instance and start the validation.

This feature also touches on some of the visual parts of Hammer, ie an option to silence the animation from the main program and make the display of the results page directly dependent on the state of the program.

#### Motivation for adding to Hammer
See [issue 61](https://github.com/health-validator/Hammer/issues/61)

### CODE REVIEW

File: Program.cs [hunk 1]
```diff
@@ -603,6 +617,69 @@ public async void StartValidation()
...
+  /// </summary>
+  public class CLIParser {
+    ParserResult<CLIOptions> cliOptions;
+
+    /// <summary>
+    /// Data storage class to store the command line options and arguments.
+    /// </summary>
+    public class CLIOptions
+    {
+      [Option('s', "scope_dir", Required = false, HelpText = "Set the scope directory")]
```
Comment (vadi2) [inline Program.cs]: Let's go with `scopedir`, easier to type.

File: Program.cs [hunk 2]
```diff
@@ -603,6 +617,69 @@ public async void StartValidation()
...
+    /// </summary>
+    public CLIParser(string[] args)
+    {
+      cliOptions = Parser.Default.ParseArguments<CLIOptions>(args);
+    }
+
+    /// <summary>
+    /// Truw whether the command line was succesfully parsed, false on error.
+    /// </summary>
+    public bool success {
```
Comment (vadi2) [inline Program.cs]: How about `parsedSuccessfully` - and then we wouldn't need a comment, it would be self-explanatory

File: Program.cs [hunk 3]
```diff
@@ -603,6 +617,69 @@ public async void StartValidation()
...
+      get {
+        var success = true;            
+        cliOptions.WithNotParsed(errors => success = false);
```
Comment (vadi2) [inline Program.cs]: Fancy... :)

File: Program.cs [hunk 4]
```diff
@@ -603,6 +617,69 @@ public async void StartValidation()
...
+        return success;
+      }
+    }
+
+    /// <summary>
+    /// Perform the actions specified by the command line.
+    /// </summary>
+    public void process()
+    {
+      cliOptions.WithParsed(result => {
```
Comment (vadi2) [inline Program.cs]:
```suggestion
      cliOptions.WithParsed(cliOptions => {
```

File: Program.cs [hunk 5]
```diff
@@ -603,6 +617,69 @@ public async void StartValidation()
...
+        AppModel.Instance.DoQmlAnimation = false;
+
+        if (result.ScopeDir != null) {
+          var scopeUri = new System.Uri(System.IO.Path.GetFullPath(result.ScopeDir));
+          AppModel.Instance.LoadScopeDirectory(scopeUri.ToString());
```
Comment (vadi2) [inline Program.cs]: Hm... I can see here how it's poorly named. Perhaps in a follow-up PR we should rename it to `SetScopeDirectory`.

File: Program.cs [hunk 6]
```diff
@@ -614,9 +691,24 @@ static int Main(string[] args)
     {
       using (var engine = new QQmlApplicationEngine())
       {
+        // We first need to register the AppModel type in QML in order to have
+        // an instance that we can work on programmatically.
         Qml.Net.Qml.RegisterType<AppModel>("appmodel");
+
+        // Now first handle the command line options to check if we should bail
```
Comment (vadi2) [inline Program.cs]:
```suggestion
        // Now we can check command line options to see if we should bail
```

File: Hammer.csproj [hunk 7]
```diff
@@ -24,6 +24,7 @@
     <PackageReference Include="Qml.Net.WindowsBinaries" Version="0.10.1" />
     <PackageReference Include="System.ValueTuple" Version="4.5.0" />
     <PackageReference Include="TextCopy" Version="1.5.2" />
+    <PackageReference Include="CommandLineParser" Version="2.5.0" />
```
[the remark this answers is not in the document]
Comment (pieter-edelman-nictiz) [reply Hammer.csproj]: That's weird, it's in the standard NuGet directory (https://www.nuget.org/packages/CommandLineParser/). Maybe it was a temporary issue?

File: Main.qml [hunk 8]
```diff
@@ -74,6 +73,10 @@ ApplicationWindow {
         width: window.width
         height: window.height - buttonsRow.height
 
+        Connections {
+            target: appmodel
+            onValidationStarted: addResourcesPage.state = "VALIDATION_RESULTS"
```
[the remark this answers is not in the document]
Comment (pieter-edelman-nictiz) [reply Main.qml]: Yeah, works for me. It's a new signal defined in Program.cs (line 26). Have you tried a clean build?

File: Main.qml [hunk 9]
```diff
@@ -199,17 +202,17 @@ ApplicationWindow {
...
-                from: "*"; to: "VALIDATED_RESOURCE"
-                NumberAnimation { property: "x"; easing.type: Easing.InBack; duration: 1000 }
+                from: "*"; to: "VALIDATION_RESULTS"
+                NumberAnimation { property: "x"; easing.type: Easing.InBack; duration: animationDuration }
             },
             Transition {
                 from: "*"; to: "ENTERING_RESOURCE"
-                NumberAnimation { property: "x"; easing.type: Easing.InBack; duration: 1000 }
-                NumberAnimation { property: "y"; easing.type: Easing.OutBack; duration: 1000 }
+                NumberAnimation { property: "x"; easing.type: Easing.InBack; duration:appmodel.animationDuration }
```
[the remark this answers is not in the document]
Comment (pieter-edelman-nictiz) [reply Main.qml]: Yup, missed this one.

File: Program.cs [hunk 10]
```diff
@@ -148,6 +150,15 @@ public bool JavaValidationCrashed
       get => _javaValidationCrashed;
       set => this.SetProperty(ref _javaValidationCrashed, value);
     }
+
+    private bool _doQmlAnimation = true;
+    /// <summary>Set to false to suppress animations in QML</summary>
```
[the remark this answers is not in the document]
Comment (pieter-edelman-nictiz) [reply Program.cs]: Yeah, I'm writing it by hand. I don't really like this approach though.

File: Program.cs (hunk 1 above)
Comment (pieter-edelman-nictiz) [reply Program.cs]: Sure 👍

File: Program.cs (hunk 2 above)
Comment (pieter-edelman-nictiz) [reply Program.cs]: Yeah, fine by me.

File: Program.cs [hunk 11]
```diff
@@ -148,6 +150,15 @@ public bool JavaValidationCrashed
       get => _javaValidationCrashed;
       set => this.SetProperty(ref _javaValidationCrashed, value);
     }
+
+    private bool _doQmlAnimation = true;
+    /// <summary>Set to false to suppress animations in QML</summary>
+    [NotifySignal]
+    public bool DoQmlAnimation {
```
[the remark this answers is not in the document]
Comment (pieter-edelman-nictiz) [reply Program.cs]: 👍

File: Program.cs (hunk 4 above)
Comment (pieter-edelman-nictiz) [reply Program.cs]: That doesn't seem right ...

File: Program.cs (hunk 5 above)
Comment (pieter-edelman-nictiz) [reply Program.cs]: Well ... it does load the scope dir, doesn't it?

**Comment by pieter-edelman-nictiz** [comment]
I updated the branch. My Github-foo is not so strong, hope I'm doing this right.

File: Hammer.csproj (hunk 7 above)
[the remark this answers is not in the document]
Comment (vadi2) [reply Hammer.csproj]: Oh there was no problem, I was saying good library pick

**Pull request merged**